### PR TITLE
add Banksy-Finance Grant Request

### PR DESCRIPTION
**Pull request by xiaopangjun**
# Grant Application Checklist

- [x] The [application template](https://github.com/w3f/Open-Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [x] A BTC or Ethereum (DAI) address for the payment of the milestones is provided inside the application.  
- [x] I have read and acknowledged the [Terms and Conditions](https://github.com/w3f/Open-Grants-Program/blob/master/src/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The total funding amount of the project is below USD $30k for initial grant applications and $100k for follow-up grants.
- [x] The initial PR contains only one commit (squash if needed before submitting your PR).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery).

### CODE REVIEW

**Comment by alxs** [comment]
Thank you for your application @xiaopangjun. Adding a link to your previous one for reference #383.

**Comment by mmagician** [comment]
@xiaopangjun Thanks for the application. I had a quick look and it indeed looks much more thought-through than the previous one. 
From what I see, M1 is covering functionalities already available in the ecosystem and brings nothing new.
As you already mention yourself, there are multiple teams building NFT solutions. Personally, I would rather support code re-use than having another implementation starting from scratch. Perhaps you would be open to exploring the existing solutions, picking the one that suits you best, forking it and building on top of it? This is just my idea, other grant members might still support your project in its current state.

Regarding your AI solution: do you have an idea of what kind of approach you want to use (neural-nets, SVM, etc)? You don't mention anything about this being an off-chain task - quite the opposite:
> We will integrate analysis modules and machine learning models and deploy them to the test chain for testing.

Does that mean you want to code the entire logic of your model in the substrate module? Are you aware that there are limits on the computation that can be executed within a block? From what I understand about most machine learning approaches, they have fairly heavy computation and/or storage requirements and I'm not sure this is suited for runtime code. Please correct me if I'm wrong and elaborate on your plan.

**Comment by xiaopangjun** [comment]
> @xiaopangjun Thanks for the application. I had a quick look and it indeed looks much more thought-through than the previous one.
>
Hello, @mmagician, Thank you very much for your affirmation of our efforts. 
>From what I see, M1 is covering functionalities already available in the ecosystem and brings nothing new.
> As you already mention yourself, there are multiple teams building NFT solutions. Personally, I would rather support code re-use than having another implementation starting from scratch. Perhaps you would be open to exploring the existing solutions, picking the one that suits you best, forking it and building on top of it? This is just my idea, other grant members might still support your project in its current state.
>
Initially, we would like to conduct our exploration for the perspective of the overall performance optimization. We agree with your opinion on code reuse very much. At present, we have not found any similar projects in Polkadot Ecology that can reuse code directly. It may be a problem with our information retrieval. Maybe you will have any better suggestions?
>Regarding your AI solution: do you have an idea of what kind of approach you want to use (neural-nets, SVM, etc)? You don't mention anything about this being an off-chain task - quite the opposite:
>
We prefer machine learning algorithms to neural network algorithms. Because machine learning algorithms are more interpretable and predictable, while neural network algorithms require heavier computation and data support. From the analysis of the test, we may choose the LightGBM or XGBOOST algorithm, which is better than the SVM algorithm, and we will select the appropriate algorithm model based on the final result.
> > We will integrate analysis modules and machine learning models and deploy them to the test chain for testing.
>
> Does that mean you want to code the entire logic of your model in the substrate module? Are you aware that there are limits on the computation that can be executed within a block? From what I understand about most machine learning approaches, they have fairly heavy computation and/or storage requirements and I'm not sure this is suited for runtime code. Please correct me if I'm wrong and elaborate on your plan.
>
So sorry to make concerns. It is our misinterpretation.  And what we want to present is to integrate the machine learning algorithm and calculate at the AI node. The data source node will provide the input. We will deploy Oracle smart contracts to the test chain. 
AI Oracle

As attached, the usER-SC Contract and oracle contract is on the blockchain, while AI Node and Data sources are under the chain. The computation and data storage of machine learning are also the factors to consider in our solution.

**Comment by mmagician** [comment]
Regarding similar solutions, you surely have done a search already because you even listed these projects in your application. For other ideas, check out [the polkadot stack](https://github.com/w3f/General-Grants-Program/blob/master/grants/polkadot_stack.md) or [the accepted applications list](https://github.com/w3f/General-Grants-Program/blob/master/grants/accepted_grant_applications.md) and do a quick search for NFTs there. I can't tell you which is best suited for your project, as you know best the requirements of your own project.

Quite frankly, I don't see a need for another NFT project, but I'll share the application with the Grants Team.

**Comment by xiaopangjun** [comment]
> Regarding similar solutions, you surely have done a search already because you even listed these projects in your application. For other ideas, check out [the polkadot stack](https://github.com/w3f/General-Grants-Program/blob/master/grants/polkadot_stack.md) or [the accepted applications list](https://github.com/w3f/General-Grants-Program/blob/master/grants/accepted_grant_applications.md) and do a quick search for NFTs there. I can't tell you which is best suited for your project, as you know best the requirements of your own project.

@mmagician Thank you very much for your guidance. We have  sorted out the NFT projects in the Polkadot ecosystem and their functions, as shown in the table below. Regarding the development of the basic functions of M1, we have decided to follow your guidance and reuse the relevant code in the projects below. The functionalities are no longer part of the delivery content of M1. We will adjust the content of M2 and M3 as the final delivery content.

| Projects | Functionalities|
-- | --
FRAME Pallet: NFTs for Substrate | NFT created, destroyed and transferred.
Usetech | NFT create, NFT Wallet and UI
RareLink | NFT create, destroyed,  p2p NFT trading
NFTT Studio | NFT Store Pallet
DNFT-Team | NFT create，NFT market，NFT auction，NFT gaming
Perpetual Altruism Ltd | NFT bridge
Starry Network | Splittable NFTs
DEGO | NFT publishing, NFT trading
Vera | p2p NFT Lending
GamePower | NFT Collectibles Wallet
Banksy Finance | NFT AI-KIT, Pool-based NFT Mortgage Lending, AI Oracle

> Quite frankly, I don't see a need for another NFT project, but I'll share the application with the Grants Team.

Banksy is not a pure NFT project. It is NFT + DeFi + AI. It is the business expansion of DeFi Lending and the exploration of the technological integration of Crypto + AI. I think it is very helpful for Polkadot's technical architecture from the perspective of DeFi and AI.

At the same time, from the perspective of NFT business, Polkadot’s current NFT project mainly focuses on NFT creation, trading, destruction, auction, cross-chain, etc. (as shown in the table above). Only one project, Vera, involves NFT Lending, but it It is in p2p mode. Banksy is a pool-based NFT mortgage lending platform. It uses AI technology to solve the valuation problem of NFT, so as to truly solve the core problem of NFT mortgage lending, which can greatly improve the liquidity of NFT, which has a great impact on Polkadot’s NFT ecosystem. 

In summary, Banksy is very different from other NFT projects, and our AI kit, NFT mortgage lending, can improve the NFT market infrastructure of the Polkadot ecology.

**Comment by semuelle** [comment]
- For a thorough originality check, you would have to have a fairly large database of files to compare it against. Where do you get this data from? If an artist submits an image for sale, which they previously published on their private blog, how do you verify it is the author selling? I just went through your marketplace and found every image I tried elsewhere.
- What is the output of the `Base Analysis`? A price? 
- Have you conducted any tests to check if you are actually able to approximate a useful market price? If so, what data did you train and test the model on? Can you point me to any previous works that suggest this might work at all?
- `Each predictor Oi has its data sources that may not overlap. Oi takes and aggregates the data from the data sources and sends its own aggregated data` -- what would stop me from running a predictor that only submits high price predictions if the author is me?

**Comment by xiaopangjun** [comment]
@semuelle Thank you for the review. Let me show the AI architecture diagram we designed, as shown in the following figure:

WX20210719-225601

> * For a thorough originality check, you would have to have a fairly large database of files to compare it against. Where do you get this data from? If an artist submits an image for sale, which they previously published on their private blog, how do you verify it is the author selling? I just went through your marketplace and found every image I tried elsewhere.

AI-KIT’s NFT originality check has not yet been implemented on Dapp. Dapp currently only implements part of the NFT functions, such as AI-Assisted NFT Creation. When you experience the Banksy marketplace, there is no originality check. This will be achieved at the next milestone. Currently the NFT originality check we designed only supports the NFTs that have been released. The data source for originality check is the NFT transaction data on each public chain. We can obtain the attribute information and transaction information of all transaction NFTs. Protect the copyright of certified NFTs and published NFTs. 

The example you gave is related to off-chain data. We will cover it in the next stage. We may use Google search engine or other means to check the similarity of works and creation time and filter out some fraud risks according to rules and policies. For the works of, usually off-chain data has a higher level of risk. As we adopt a risk-based pricing strategy, for works with higher risk levels, we may refuse to lend or adopt lower pricing amounts and higher repayment rates. The Banksy platform has a complete whitelist mechanism. If the author himself submits for sale and is not listed on the mainstream trading platform, he can apply to the platform and submit relevant certification materials for review. After the review is passed, he will be added to the whitelist.

> * What is the output of the `Base Analysis`? A price?

In Base Analysis, we will clean the collected NFT artwork data, complete the construction of the index system, process the basic information of each artwork, build buyer portraits and seller portraits, and provide data support for the machine learning model. In the AI model construction phase, we will build multiple models, including the pre-loan asset evaluation model, original evaluation model, mortgage fraud early warning model, post-loan asset monitoring model, etc. Each model has a corresponding output, and finally our comprehensive model Output and mortgage strategy, output collateral risk level, pricing, loan interest rate, and other content.

> * Have you conducted any tests to check if you are actually able to approximate a useful market price? If so, what data did you train and test the model on? Can you point me to any previous works that suggest this might work at all?

Our team has experienced lending risk control experts in the banking industry and has experienced algorithmic experience in asset pricing. In banking, we conduct business based on the borrower’s basic information, user transaction flow, collateral market prices, and collateral historical transaction information. The technology of data analysis and modeling has been very mature and has been widely used in bank lending risk control systems. We will build multiple AI models to complete our risk control system. Different models have different responsibilities, and the choice of training data and test data will be different. For example pre-loan risk assessment model, we will divide the risk level (high, medium, and low) according to the set risk control strategy (the rise and fall of assets in the next month), and create risk level labels based on transaction history data, and keep the most recent one or more The monthly data is used as a test to test the effect of the risk control model through backtesting, and use the recall rate, accuracy rate, F1 and other indicators to evaluate the model.

Our NFT mortgage lending currently supports the mainstream NFTs with relatively large transaction volume, that is, the whitelist mechanism, so that there are enough historical transaction data, NFT attribute data, and social data as the information input for model training. The data dimension of NFT is enough to make the trained model more robust and generalization ability stronger. We can provide fuller proof of this in the milestone work.

> * `Each predictor Oi has its data sources that may not overlap. Oi takes and aggregates the data from the data sources and sends its own aggregated data` -- what would stop me from running a predictor that only submits high price predictions if the author is me?

The following four methods can effectively prevent a predictor that only submits high-priced predictions.

1. The on-chain aggregation function uses the median, so a small number of abnormal predictors will not affect the results.
2. Ensure that the AI prediction predictor will not be tampered with through technologies such as software integrity checking
3. To run the AI predictor, you need to apply for certification through the community, which has higher requirements.
4. The verification service conducts post-checks on the results submitted to the chain, especially for high-value transactions, and compares the results with data directly from authoritative data sources. When abnormal data is found, the AI predictor will be removed.

**Comment by semuelle** [comment]
Thanks for the follow-up, @xiaopangjun. I am familiar with the models used for risk assessment in the banking industry, perhaps that is why I have my doubts that this would work. I also find an "AI NFT price oracle pallet" too specialised to be useful for the greater ecosystem and, as @mmagician pointed out, there are already pallets for NFT creation and auctions. That being said, I'd be happy to give my approval for an application with the third milestone only.

**Comment by xiaopangjun** [comment]
Thank you for your feedback @semuelle.

> I am familiar with the models used for risk assessment in the banking industry, perhaps that is why I have my doubts that this would work. I also find an "AI NFT price oracle pallet" too specialised to be useful for the greater ecosystem

We will continue to advance the work of M1 and M2.

> as [@mmagician](https://github.com/mmagician) pointed out, there are already pallets for NFT creation and auctions. That being said, I'd be happy to give my approval for an application with the third milestone only.

Thank you  for your support to M3, which is our core work. We will remove the contents of M1 and M2 as required and keep the grant for M3.

**Comment by semuelle** [comment]
Hi @xiaopangjun, thanks for the update. Could you add some technical specs to the milestone? Is it all smart contracts? Which parts of the architecture diagrams does it encompass? The application now is very general for the one milestone.

**Comment by xiaopangjun** [comment]
> Hi @xiaopangjun, thanks for the update. Could you add some technical specs to the milestone? Is it all smart contracts? Which parts of the architecture diagrams does it encompass? The application now is very general for the one milestone.

@semuelle thanks for your feedback. We have added more technical content and updated the milestones according to your requirements, please check it out. If there are other issues with the milestone, please point it out again and we will update it as needed.

**Review by semuelle — APPROVED** [review]
Thanks for the update.

**Review by mmagician — APPROVED** [review]
Alright why not, the revised scope looks more acceptable!

**Review by Noc2 — APPROVED** [review]

**Comment by semuelle** [comment]
@xiaopangjun, I see that the name of your project/company has changed. Could you amend the contract to reflect this (team name, github links, etc.)? Also, could you provide a brief update on your work? It seems that your `banksea-polkadot` repo has been inactive in a while.

**Comment by alxs** [comment]
@xiaopangjun could you give us a quick update on your progress? Note that since the first milestone is already very delayed and we already asked you for an update, if we don't hear from you in the next 2 weeks we will assume you're no longer interested and terminate the agreement.
